### PR TITLE
Make ARCore plugin work

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Working on making this plugin work.

You need to compile Godot using it's 3.x branch with this PR included: https://github.com/godotengine/godot/pull/47455 (or use the branch on my repo for that pull request). Be sure to also build Android builds and copy the resulting files into your templates folder.

Then make sure to create a new `api.json` by running: `godot --gdnative-generate-json-api api.json`

Copy the resulting `api.json` file into `plugin/libs/godot-cpp/godot-headers` overwriting the file that is already there. Also copy the contents of `modules/gdnative/include` from your Godot build into this folder

Now you can run our python script to generate all the godot-cpp classes:
```
cd plugin/libs/godot-cpp
python ../../../generate.py
```

Now you can build the plugin itself by calling `./gradlew :generatePluginBinary` and follow the rest of the instructions in the readme for creating a custom android build.

Currently I'm investigating why the plugin doesn't load.

*NOTE* CI will currently fail as godot-headers has to be updated in the godot-cpp repo once PR 47455 has been merged..

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
@m4gr3d , ok for now I've put our binding layer directly in here, best to get it to work here and tested and then look into adding it to gdnative itself as it may take some time to get that merged.

Everything compiles but I can't test it yet, for some reason I can't compile:
https://github.com/godotengine/godot/pull/41026
There is one change I have added ontop of your PR, I'll probably do a PR for that on your repo. All it does is expose `set_interface` to GDNative so we can call it.
Hence line 658 in arcore_interface.cpp being commented out for now. 

Without that however, nothing will actually work :)

**Comment by rcoppy** [comment]
Hi there! @BastiaanOlij Just wondering, is this repo currently dead? Interested in using ARCore with Godot for a work project. Thanks for your time!

**Comment by BastiaanOlij** [comment]
Hey @rcoppy ,

Hybernating, not dead :) I've been wanting to get back to this for ages but too many competing projects at the moment. Plus my phone broke...
Following the instructions on my post on the 7th of November it would be good if others could give it a swing and see how far they get. 

I actually just received a new Android phone which is on the list of ARCore supported devices so should be able to start doing some testing again soon

**Comment by rcoppy** [comment]
Hey @BastiaanOlij ! So sorry I never got back to you--github notifications can be strange sometimes! I'm a bit busy too at the moment but if I open up I'll definitely see what I can do 🙂

**Comment by clemens-tolboom** [comment]
> Following the instructions on my post on the 7th of November

Where is this blog located :-)

**Comment by BastiaanOlij** [comment]
> 
> 
> > Following the instructions on my post on the 7th of November
> 
> Where is this blog located :-)

Not a blog, just instructions a few posts up: https://github.com/GodotVR/godot_arcore/pull/6#issuecomment-723435087

**Comment by BastiaanOlij** [comment]
After months I've finally found some time to revisit this and rebased the plugin fixing a bunch of issues. It compiles now, still doesn't run, see blocking issues in the OP that need to be solved upstream first.

**Comment by rcoppy** [comment]
Awesome!

**Comment by Railwanderer** [comment]
Hey @BastiaanOlij  thank you so much for your effort this far already! I tried out VR in Godot and it worked very good for my purpose. Now I wondered if AR was also possible and followed along all the preceding posts to find the work is still kind of going. I wish I could help somehow to push this forward. Unfortunately I don't know what is missing and what is there already. I am really looking forward to use it in a project!

**Comment by BastiaanOlij** [comment]
@Railwanderer I wish I had more time to pursue this as I feel we're very close to something working. The biggest issue right now is just in the order in which things are initialized.

**Comment by Railwanderer** [comment]
@BastiaanOlij thanks for the reply! I will try to stick around and see if I can be useful.

**Comment by BastiaanOlij** [comment]
I'm currently having an issue with the library not loading and I'm not sure why. I did figure out that because we're using an 'aar' and our ARCore object is a singleton, there are no GDNative objects, that Godot removes the `.gdnlib` file from the `project.godot` file resulting the Godot not attempting to load our library.

I've worked around that for now by introducing a `ARCoreConfig` object that doesn't do anything but triggers the load. Unfortunately it can't load it:
```
01-05 00:11:32.938  5730  9557 I godot   : Godot Engine v3.5.beta.custom_build.97b0d2ef3 - https://godotengine.org
01-05 00:11:32.943  5730  9557 I godot   : OpenGL ES 3.0 Renderer: Mali-G78
01-05 00:11:32.943  5730  9557 I godot   : Async. shader compilation: OFF
01-05 00:11:32.956  5730  9557 I godot   :
01-05 00:11:33.164  5730  9557 E godot   : ERROR: Can't open dynamic library: libgodot_arcore.so, error: dlopen failed: library "libgodot_arcore.so" not found.
01-05 00:11:33.164  5730  9557 E godot   :    at: open_dynamic_library (platform\android\os_android.cpp:233) - Condition "!p_library_handle" is true. Returned: ERR_CANT_OPEN
01-05 00:11:33.164  5730  9557 E godot   : ERROR: No valid library handle, can't get symbol from GDNative object
01-05 00:11:33.164  5730  9557 E godot   :    at: get_symbol (modules\gdnative\gdnative.cpp:510) - No valid library handle, can't get symbol from GDNative object
01-05 00:11:33.164  5730  9557 E godot   : ERROR: No nativescript_init in "res://addons/godot-arcore/lib/arm64-v8a/libgodot_arcore.so" found
01-05 00:11:33.164  5730  9557 E godot   :    at: init_library (modules\gdnative\nativescript\nativescript.cpp:1503) - No nativescript_init in "res://addons/godot-arcore/lib/arm64-v8a/libgodot_arcore.so" found
```

The so file is indeed not in that location but it is here:
image
But I figured it would remap them? I could add remap entries manually I maybe...

@m4gr3d do you have any ideas what is going wrong here?

**Comment by BastiaanOlij** [comment]
@m4gr3d I've attached a more complete log for the ARCore run, I can see that the ARCore plugin is loaded which I guess is triggered by the `.gdap` file, but it is not initialized. Can you see anything in the log that tells us why ?

[arcore log.txt](https://github.com/GodotVR/godot_arcore/files/7811595/arcore.log.txt)

**Comment by m4gr3d** [comment]
@BastiaanOlij `ARCorePlugin` should be loading the shared library as well as any dependencies it may have. Here is a reference for what the code should look like: https://github.com/godotengine/godot/blob/master/platform/android/java/lib/src/org/godotengine/godot/GodotLib.java#L45.

You also need to include in `ARCorePlugin` the relative path to the `gdnlib` file. Here's an example (in kotlin, but the same concept applies): https://github.com/GodotVR/godot_openxr/blob/master/android/src/main/java/org/godotengine/plugin/vr/openxr/OpenXRPlugin.kt#L72

**Comment by BastiaanOlij** [comment]
> @BastiaanOlij `ARCorePlugin` should be loading the shared library as well as any dependencies it may have. Here is a reference for what the code should look like: https://github.com/godotengine/godot/blob/master/platform/android/java/lib/src/org/godotengine/godot/GodotLib.java#L45.

Not sure I follow this one, that is code on the Godot side so that should be there? The godot lib is included here:

https://github.com/GodotVR/godot_arcore/tree/make_this_work/plugin/libs/godot-lib.debug.aar

And as far as I can see it's included here: https://github.com/GodotVR/godot_arcore/blob/make_this_work/plugin/build.gradle#L59

Do I need to do something to trigger it to load?

> You also need to include in `ARCorePlugin` the relative path to the `gdnlib` file. Here's an example (in kotlin, but the same concept applies): https://github.com/GodotVR/godot_openxr/blob/master/android/src/main/java/org/godotengine/plugin/vr/openxr/OpenXRPlugin.kt#L72

I've added the (hopefully correct) JAVA code here: https://github.com/GodotVR/godot_arcore/blob/make_this_work/plugin/src/main/java/org/godotengine/plugin/arcore/ARCorePlugin.java#L34

Should I attempt converting the whole thing to kotlin? I've never used kotlin before so that'll be fun :) Seeing it used to load in fine I'm hoping we can get it to work as is.

Obviously seeing OpenXR loads fine and is a singleton, ARCore has to work too, I should be able to remove the unnecesairy config object.

File: plugin/src/main/java/org/godotengine/plugin/arcore/ARCorePlugin.java [hunk 1]
```diff
@@ -28,4 +29,9 @@ public View onMainCreate(Activity activity) {
 	public String getPluginName() {
 		return "ARCorePlugin";
 	}
+
+	@Override
+	public Set<String> getPluginGDNativeLibrariesPaths() {
+		return Set.of("addons/godot-openxr/godot_arcore.gdnlib");
```
Comment (dasg34) [inline plugin/src/main/java/org/godotengine/plugin/arcore/ARCorePlugin.java:35]:
```suggestion
		return Set.of("addons/godot_arcore/godot_arcore.gdnlib");
```

File: plugin/src/main/cpp/gdnative/godot_arcore.cpp [hunk 2]
```diff
@@ -1,34 +1,226 @@
...
+	godot_vector2 ret_size;
+	api->godot_vector2_new(&ret_size, arvr_size.x, arvr_size.y);
+
+	return ret_size;
+};
+
+godot_transform godot_arvr_get_transform_for_eye(void *p_data, godot_int p_eye, godot_transform *p_cam_transform) {
+	ARCoreInterface *data = (ARCoreInterface *)p_data;
+
+	godot::Transform *cam_transform = (godot::Transform *)&p_cam_transform;
```
Comment (dasg34) [inline plugin/src/main/cpp/gdnative/godot_arcore.cpp:173]: `p_cam_transform` is already pointer of `godot_Transform`. i don't think we need to dereference it again.
```suggestion
	godot::Transform *cam_transform = (godot::Transform *)p_cam_transform;
```

**Review by dasg34 — COMMENTED** [review]
Hi, i found some typos.

**Comment by BastiaanOlij** [comment]
Thanks @dasg34 , I had found a few of those already but these I missed!:)

File: plugin/src/main/java/org/godotengine/plugin/arcore/ARCorePlugin.java (hunk 1 above)
Comment (BastiaanOlij) [reply plugin/src/main/java/org/godotengine/plugin/arcore/ARCorePlugin.java]: Thanks!

File: plugin/src/main/cpp/gdnative/godot_arcore.cpp (hunk 2 above)
Comment (BastiaanOlij) [reply plugin/src/main/cpp/gdnative/godot_arcore.cpp]: Thanks:)

**Comment by BastiaanOlij** [comment]
Ok bunch of things fixed up but its still not loading properly. Must still be missing something...

**Comment by BastiaanOlij** [comment]
owh I'm stupid!

**Comment by BastiaanOlij** [comment]
Ok.. so the crash was because I can't call `godot::print` before gdnative is initialised, so replaced a few with `ALOGV` :)

Now it crashes on creating the ARVRInterface.. ok progress, to be continued tomorrow.

**Comment by BastiaanOlij** [comment]
OK..so `godot_arcore_nativescript_init` isn't being called at all, presumably because we have no GDNS file. This is annoying..

**Comment by BastiaanOlij** [comment]
Alright, looks like initialization now works :) It sets up and registers our ARCoreInterface properly now and gets all the way to creating the ARCore session. That fails presumably because ARCore isn't installed. That likely means our setup activity isn't run. Getting real close now I think...

**Comment by BastiaanOlij** [comment]
Seeing I don't have time to actively work on this and this does contain a whole bunch of fixes, I'm going to merge this so others can work on it further.

**Pull request merged**